### PR TITLE
add e2e test for deployment replicas syncer

### PR DESCRIPTION
**Pull request by chaosi-zju**
**What type of PR is this?**

/kind failing-test

**What this PR does / why we need it**:

add e2e test for deployment replicas syncer

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by Vacant2333** [comment]
hi @chaosi-zju , is the pr ready for review now?

**Comment by chaosi-zju** [comment]
> hi @chaosi-zju , is the pr ready for review now?

Hi @Vacant2333, it is ready~

Thanks!

File: test/e2e/deploymentreplicassyncer_test.go [hunk 1]
```diff
@@ -30,81 +31,196 @@ import (
...
 
+	ginkgo.BeforeEach(func() {
+		namespace = testNamespace
+		deploymentName = deploymentNamePrefix + rand.String(RandomStrLength)
+		hpaName = deploymentName
+		policyName = deploymentName
+		bindingName = names.GenerateBindingName(util.DeploymentKind, deploymentName)
+
+		// sort member clusters in increasing order
+		targetClusters = framework.ClusterNames()[0:2]
```
Comment (chaosi-zju) [inline test/e2e/deploymentreplicassyncer_test.go:57]: notes: now it is no need to sort, I will remove this line later

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks @chaosi-zju @Vacant2333 
/approve

**Comment by XiShanYongYe-Chang** [comment]
Hi @chaosi-zju, the newly added e2e seems have the cccasional error:

```log
• [FAILED] [430.089 seconds]
deployment replicas syncer testing when policy is Divided schedule type, remove one cluster's replicas [It] 0/1 case, manually modify replicas from 2 to 1
/home/runner/work/karmada/karmada/test/e2e/deploymentreplicassyncer_test.go:151

  Captured StdOut/StdErr Output >>
  I0319 02:40:22.362638   51005 deployment.go:75] Waiting for deployment(karmadatest-78jb4/deploy-wzfcb) synced on cluster(member1)
  I0319 02:40:27.369969   51005 deploymentreplicassyncer_test.go:210] in member1 cluster, got: 1, expect: 1
  I0319 02:40:27.370051   51005 deployment.go:75] Waiting for deployment(karmadatest-78jb4/deploy-wzfcb) synced on cluster(member2)
  I0319 02:47:27.407015   51005 deployment.go:123] Waiting for deployment(karmadatest-78jb4/deploy-wzfcb) disappears on cluster(member1)
  I0319 02:47:32.431213   51005 deployment.go:123] Waiting for deployment(karmadatest-78jb4/deploy-wzfcb) disappears on cluster(member2)
  << Captured StdOut/StdErr Output

  Timeline >>
  STEP: Creating PropagationPolicy(karmadatest-78jb4/deploy-wzfcb) @ 03/19/24 02:40:22.344
  STEP: Creating Deployment(karmadatest-78jb4/deploy-wzfcb) @ 03/19/24 02:40:22.352
  STEP: Creating HPA(karmadatest-78jb4/deploy-wzfcb) @ 03/19/24 02:40:22.356
  STEP: step1: propagate 2 replicas to two clusters @ 03/19/24 02:40:22.362
  STEP: Waiting for deployment(karmadatest-78jb4/deploy-wzfcb) synced on member clusters @ 03/19/24 02:40:22.362
  STEP: Waiting for deployment(karmadatest-78jb4/deploy-wzfcb) synced on member clusters @ 03/19/24 02:40:27.37
  [FAILED] in [It] - /home/runner/work/karmada/karmada/test/e2e/framework/deployment.go:82 @ 03/19/24 02:47:27.371
  STEP: Removing PropagationPolicy(karmadatest-78jb4/deploy-wzfcb) @ 03/19/24 02:47:27.371
  STEP: Removing Deployment(karmadatest-78jb4/deploy-wzfcb) @ 03/19/24 02:47:27.393
  STEP: Removing HPA(karmadatest-78jb4/deploy-wzfcb) @ 03/19/24 02:47:27.396
  STEP: Check if deployment(karmadatest-78jb4/deploy-wzfcb) disappears on member clusters @ 03/19/24 02:47:27.406
  << Timeline

  [FAILED] Timed out after 420.000s.
  Expected
      <bool>: false
  to equal
      <bool>: true
  In [It] at: /home/runner/work/karmada/karmada/test/e2e/framework/deployment.go:82 @ 03/19/24 02:47:27.371

  Full Stack Trace
    github.com/karmada-io/karmada/test/e2e/framework.WaitDeploymentPresentOnClusterFitWith({0xc000036f19, 0x7}, {0xc000814e88, 0x11}, {0xc0008cc094, 0xc}, 0xc000ae1290)
    	/home/runner/work/karmada/karmada/test/e2e/framework/deployment.go:82 +0x614
    github.com/karmada-io/karmada/test/e2e/framework.WaitDeploymentPresentOnClustersFitWith.func1()
    	/home/runner/work/karmada/karmada/test/e2e/framework/deployment.go:100 +0xc5
    github.com/karmada-io/karmada/test/e2e/framework.WaitDeploymentPresentOnClustersFitWith({0xc00083d940, 0x1, 0x1}, {0xc000814e88, 0x11}, {0xc0008cc094, 0xc}, 0xc000ae1290)
    	/home/runner/work/karmada/karmada/test/e2e/framework/deployment.go:98 +0x236
    github.com/karmada-io/karmada/test/e2e.assertDeploymentWorkloadReplicas({0xc000814e88, 0x11}, {0xc0008cc094, 0xc}, {0xc000991100, 0x2, 0x10308ec?}, {0xc000227a68, 0x2, 0x2})
    	/home/runner/work/karmada/karmada/test/e2e/deploymentreplicassyncer_test.go:209 +0x173
    github.com/karmada-io/karmada/test/e2e.glob..func11.5.2.1()
    	/home/runner/work/karmada/karmada/test/e2e/deploymentreplicassyncer_test.go:153 +0x11b
    github.com/karmada-io/karmada/test/e2e.glob..func11.5.2()
    	/home/runner/work/karmada/karmada/test/e2e/deploymentreplicassyncer_test.go:152 +0xed
```

It run failed on the master branch: https://github.com/karmada-io/karmada/actions/runs/8336573479/job/22814077291

**Comment by chaosi-zju** [comment]
> the newly added e2e seems have the cccasional error

the problem found and fixed in #4729

**Pull request merged**